AdyenEncryption/JOSE/Encryption Algorithms/Content Encryption Algorithms/JWAA256CBCHS512Algorithm.swift

```diff
@@ -0,0 +1,76 @@
+//
+// Copyright (c) 2021 Adyen N.V.
+//
+// This file is open source and available under the MIT license. See the LICENSE file for more info.
+//
+
+import CommonCrypto
+import Foundation
+
+internal struct JWAA256CBCHS512Algorithm: JWAEncryptionAlgorithm {
+    internal let keyLength: Int = kCCKeySizeAES256 * 2
+    
+    internal let initializationVectorLength: Int = kCCBlockSizeAES128
+    
+    internal let name: String = "A256CBC-HS512"
+    
+    internal func encrypt(input: JWAInput) throws -> JWAOutput {
+        guard input.key.count == keyLength else {
+            throw EncryptionError.invalidKey
+        }
+        guard input.initializationVector.count == initializationVectorLength else {
+            throw EncryptionError.invalidInitializationVector
+        }
+
+        let (aesKey, hmacKey) = deriveKeys(from: input.key)
+        
+        let operation = CCOperation(kCCEncrypt)
+        let option = CCOptions(kCCOptionPKCS7Padding)
+        let encryptedPayload = try aesCrypt(operation: operation,
+                                            options: option,
+                                            keyData: aesKey,
+                                            initializationVector: input.initializationVector,
+                                            dataIn: input.payload)
+        let authenticationTag = try createAuthenticationTag(withAdditionalAuthenticationData: input.additionalAuthenticationData,
+                                                            initializationVector: input.initializationVector,
+                                                            encryptedPayload: encryptedPayload,
+                                                            hmacKey: hmacKey)
+        return JWAOutput(encryptedPayload: encryptedPayload, authenticationTag: authenticationTag)
+    }
+    
+    private func deriveKeys(from key: Data) -> (aesKey: Data, hmacKey: Data) {
+        let subKeyLength = keyLength / 2
+        
+        let keyStart: Data.Index = key.startIndex
+        let keyMiddle: Data.Index = keyStart.advanced(by: subKeyLength)
+        let keyEnd: Data.Index = keyMiddle.advanced(by: subKeyLength)
+        
+        let hmacKey = key.subdata(in: keyStart..<keyMiddle)
+        let aesKey = key.subdata(in: keyMiddle..<keyEnd)
+        
+        return (aesKey, hmacKey)
+    }
+    
+    private func createAuthenticationTag(withAdditionalAuthenticationData: Data,
+                                         initializationVector: Data,
+                                         encryptedPayload: Data,
+                                         hmacKey: Data) throws -> Data {
+        var hmacData = Data()
```

Comment:
this is a reusable algorithm, I think this way is better.

Reply:
sorry, maybe i did not explain it good. i meant we can do this instead of creating mutable data and appending blocks. this is probably better from memory performance standpoint.
```swift
        let hmacData = Data([
            withAdditionalAuthenticationData,
            initializationVector,
            encryptedPayload,
            tagLengthData(forAdditionalAuthenticationData: withAdditionalAuthenticationData)
        ].flatMap { $0 })
```

Reply:
![Screen Shot 2021-08-30 at 11 50 48 AM](https://user-images.githubusercontent.com/2619933/131321426-ec0f1a95-5bfd-4c11-938e-8edd5edf520b.png)
![Screen Shot 2021-08-30 at 11 50 57 AM](https://user-images.githubusercontent.com/2619933/131321483-e9123c14-2e2b-4d05-b0bd-db6c820ffb8c.png)
There is no difference in performance, and if there was its imperceptible by human beings :-D